compiler/rustc_errors/src/lib.rs

```diff
@@ -1152,31 +1158,33 @@ impl HandlerInner {
         FatalError
     }
 
-    fn err(&mut self, msg: &str) {
-        self.emit_error(Error { lint: false }, msg);
+    fn err(&mut self, msg: &str) -> ErrorReported {
+        self.emit_error(Error { lint: false }, msg)
     }
 
     /// Emit an error; level should be `Error` or `Fatal`.
-    fn emit_error(&mut self, level: Level, msg: &str) {
+    fn emit_error(&mut self, level: Level, msg: &str) -> ErrorReported {
         if self.treat_err_as_bug() {
             self.bug(msg);
         }
-        self.emit_diagnostic(&Diagnostic::new(level, msg));
+        self.emit_diagnostic(&Diagnostic::new(level, msg))
     }
 
     fn bug(&mut self, msg: &str) -> ! {
         self.emit_diagnostic(&Diagnostic::new(Bug, msg));
         panic::panic_any(ExplicitBug);
     }
 
-    fn delay_as_bug(&mut self, diagnostic: Diagnostic) {
+    fn delay_as_bug(&mut self, diagnostic: Diagnostic) -> Option<ErrorReported> {
         if self.quiet {
-            return;
+            return None;
         }
```

Comment:
Why does a `quiet` flag exist? This doesn't seem sound for `delay_span_bug` and fatal errors in general. I'll go looking in the git history.

Reply:
`quiet` is from https://github.com/rust-lang/rust/commit/5cf300d695c4ac6e4bdab8fe5c48de6b05b2cd96 from #84176.

@GuillaumeGomez do you know if that change can be reverted? It shouldn't have been merged, it makes it look like it's legitimate to disable diagnostics (to be clear, this is very dangerous, especially as more things rely on `delay_span_bug` and error emission).

There is no way to keep `rustdoc` doing what that PR does and still e.g. create `ty::Error`. An error *has to* be emitted to have `ty::Error`.

I don't think compromising the integrity of the rest of the system is worth getting access to method call annotations in code, which seems to be the only use of this misfeature.

AFAICT the only way to get the information you want is to run a *separate* compilation pass, and sinkhole errors at the actual point where they would be printed to `stderr`, while letting the compiler properly track the fact that errors happened, and then scavenge whatever data is available, despite those errors.

Reply:
If you can provide a way to get this information without emitting errors, sure. But please provide it before reverting this one.